components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/token/AbstractJWTGenerator.java

```diff
@@ -0,0 +1,633 @@
+    }
+
+    private String buildJWTBody(Map<String, String> claims) {
+
+        StringBuilder jwtBuilder = new StringBuilder();
+        jwtBuilder.append("{");
+        if (claims != null) {
+            Iterator<Map.Entry<String, String>> entryIterator = claims.entrySet().iterator();
+            while (entryIterator.hasNext()) {
+                Map.Entry<String, String> entry = entryIterator.next();
+                String key = entry.getKey();
+                if("exp".equals(key) || "nbf".equals(key) || "iat".equals(key)) {
+                    //These values should be numbers.
+                    jwtBuilder.append("\"" + key + "\":" + entry.getValue() + ",");
+                }
+                else {
+                    jwtBuilder.append("\"" + key + "\":\"" + entry.getValue() + "\",");
+                }
+            }
+        }
+
+        if (jwtBuilder.length() > 1) {
+            jwtBuilder.delete(jwtBuilder.length() - 1, jwtBuilder.length());
+        }
+
+        jwtBuilder.append("}");
+        return jwtBuilder.toString();
+    }
+
+    public abstract Map<String, String> populateStandardClaims(Map<String, Object> saml2Assertions)
+            throws AppManagementException ;
+    public abstract Map<String, String> populateCustomClaims(Map<String, Object> saml2Assertions)
+            throws AppManagementException ;
+    
+
+    public ClaimsRetriever getClaimsRetriever() {
+        return claimsRetriever;
+    }
+
+    /**
+     * Helper method to sign the JWT
+     *
+     * @param assertion
+     * @param endUserName
+     * @return signed assertion
+     * @throws org.wso2.carbon.appmgt.api.AppManagementException
+     */
+    private byte[] signJWT(String assertion, String endUserName)
+            throws AppManagementException {
+
+        try {
+            String tenantDomain = MultitenantUtils.getTenantDomain(endUserName);
+            int tenantId = getTenantId(endUserName);
+
+            Key privateKey = null;
+
+            if (!(privateKeys.containsKey(tenantId))) {
+                AppManagerUtil.loadTenantRegistry(tenantId);
+                //get tenant's key store manager
+                KeyStoreManager tenantKSM = KeyStoreManager.getInstance(tenantId);
+
+                if(!tenantDomain.equals(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME)){
+                    //derive key store name
+                    String ksName = tenantDomain.trim().replace(".", "-");
+                    String jksName = ksName + ".jks";
+                    //obtain private key
+                    //TODO: maintain a hash map with tenants' private keys after first initialization
+                    privateKey = tenantKSM.getPrivateKey(jksName, tenantDomain);
+                }else{
+                    try{
+                        privateKey = tenantKSM.getDefaultPrivateKey();
+                    }catch (Exception e){
+                        log.error("Error while obtaining private key for super tenant",e);
```

Comment:
Fail the execution, as there is no point continuing when we can not proceed without signing.